server/users.go

```diff
@@ -1,317 +1,153 @@
 package server
 
 import (
-	"context"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"net/http"
-	"net/url"
 
 	"github.com/bouk/httprouter"
 	"github.com/influxdata/chronograf"
 )
 
-// NewSourceUser adds user to source
-func (h *Service) NewSourceUser(w http.ResponseWriter, r *http.Request) {
-	var req userRequest
-	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
-		invalidJSON(w, h.Logger)
-		return
-	}
-
-	if err := req.ValidCreate(); err != nil {
-		invalidData(w, err, h.Logger)
-		return
-	}
-
-	ctx := r.Context()
-	srcID, ts, err := h.sourcesSeries(ctx, w, r)
-	if err != nil {
-		return
-	}
-
-	store := ts.Users(ctx)
-	user := &chronograf.User{
-		Name:        req.Username,
-		Passwd:      req.Password,
-		Permissions: req.Permissions,
-		Roles:       req.Roles,
-	}
-
-	res, err := store.Add(ctx, user)
-	if err != nil {
-		Error(w, http.StatusBadRequest, err.Error(), h.Logger)
-		return
-	}
-
-	if err != nil {
-		Error(w, http.StatusBadRequest, err.Error(), h.Logger)
-		return
-	}
-
-	su := newUserResponse(srcID, res.Name).WithPermissions(res.Permissions)
-	if _, hasRoles := h.hasRoles(ctx, ts); hasRoles {
-		su.WithRoles(srcID, res.Roles)
-	}
-	w.Header().Add("Location", su.Links.Self)
-	encodeJSON(w, http.StatusCreated, su, h.Logger)
+type userRequest struct {
+	ID       uint64 `json:"id"`
+	Name     string `json:"name"`
+	Provider string `json:"provider,omitempty"`
+	Scheme   string `json:"scheme,omitempty"`
 }
 
-// SourceUsers retrieves all users from source.
-func (h *Service) SourceUsers(w http.ResponseWriter, r *http.Request) {
-	ctx := r.Context()
-	srcID, ts, err := h.sourcesSeries(ctx, w, r)
-	if err != nil {
-		return
+func (r *userRequest) ValidCreate() error {
+	if r.Name == "" {
+		return errors.New("Name required on Chronograf User request body")
 	}
-
-	store := ts.Users(ctx)
-	users, err := store.All(ctx)
-	if err != nil {
-		Error(w, http.StatusBadRequest, err.Error(), h.Logger)
-		return
+	if r.Provider == "" {
+		return errors.New("Provider required on Chronograf User request body")
 	}
-
-	_, hasRoles := h.hasRoles(ctx, ts)
-	ur := make([]userResponse, len(users))
-	for i, u := range users {
-		usr := newUserResponse(srcID, u.Name).WithPermissions(u.Permissions)
-		if hasRoles {
-			usr.WithRoles(srcID, u.Roles)
-		}
-		ur[i] = *usr
+	if r.Scheme == "" {
+		return errors.New("Scheme required on Chronograf User request body")
 	}
+	return nil
+}
 
-	res := usersResponse{
-		Users: ur,
+func (r *userRequest) ValidUpdate() error {
+	if r.Name == "" && r.Provider == "" && r.Scheme == "" {
+		return errors.New("No fields to update")
```

Comment:
What do you think about giving a more informative message to the front-end developers like:
"No fields to update.  name, provider, or scheme required"

Reply:
What fields are allowed to be updated will probably change, so we'll leave this for now but add a comment in the code.